Make the magic support for-else, while-else, try-else, try-except*, and match

whitphx opened the pull request:
## Describe your changes

## GitHub Issue Link (if applicable)
#9109

## Testing Plan
- Unit Tests (Python)

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

whitphx committed:
Make the magic support for-else, while-else, try-else, try-except*, and match

whitphx commented on lib/streamlit/runtime/scriptrunner/magic.py:
```diff
@@ -119,8 +137,6 @@ def _modify_ast_subtree(
 
     ast.fix_missing_locations(tree)
 
-    return tree
```
`_modify_ast_subtree` is used in many places expecting it changes the passed object in-place, so I removed the returned value as it's rather confusing.
There were some lines using the returned value, so I modified them as well.

whitphx committed:
Skip the unit tests conditionally

kajarenc commented on lib/streamlit/runtime/scriptrunner/magic.py:
```diff
@@ -15,10 +15,14 @@
 from streamlit import config
 
+MATCH_SYNTAX_AVAILABLE = sys.version_info >= (3, 10)
+TRY_STAR_SYNTAX_AVAILABLE = sys.version_info >= (3, 11)
```
minor thing: maybe change the `TRY_STAR_SYNTAX_AVAILABLE` constant name to `EXCEPT_STAR_SYNTAX_AVAILABLE` to be aligned with the statement name.

whitphx replied:
Thanks!
It makes sense, but fyi I named it after the `ast.TryStar` object.
Also, if we focus on the availability of the specific feature, `EXCEPTION_GROUP_SYNTAX_AVAILABLE` might be another option.
What do you thing is the best? All work for me.

kajarenc replied:
Yes, both options work for me; I think it can remain as is. 👍

whitphx committed:
Merge branch 'develop' into magic-support-modern-sytaxes

kajarenc committed:
make mypy happy

Mypy not smart enough to infer information about version from constant, but doing check directly resolves false positive error

kajarenc committed:
fix failing tests because of constant removal

whitphx commented on lib/tests/streamlit/runtime/scriptrunner/magic_test.py:
```diff
@@ -142,7 +143,7 @@ def test_try_star_statement(self):
 
     @unittest.skipIf(
-        not magic.TRY_STAR_SYNTAX_AVAILABLE, "Not supported in this Python version"
+        not sys.version_info >= (3, 11), "Not supported in this Python version"
```
Sorry the original code here was wrong where `TRY_STAR_SYNTAX_AVAILABLE` should be `MATCH_SYNTAX_AVAILABLE`, and now it should be `sys.version_info >= (3, 10)`.

kajarenc replied:
O, oki oki

I am fixing mypy errors and, and probably the best way it get rid of constants and do checks inline (mypy not smart enough to infer from constant, but could do that from an inline statement)

whitphx committed:
Fix the unit test skip condition

kajarenc committed:
fix typo

vdonato reviewed (approved):
LGTM -- thanks @whitphx!
